local-cli/server/util/attachHMRServer.js

```diff
@@ -16,11 +16,254 @@ const url = require('url');
  * Hot Module Replacement updates to the simulator.
  */
 function attachHMRServer({httpServer, path, packagerServer}) {
-  let client = null;
 
-  function disconnect() {
-    client = null;
-    packagerServer.setHMRFileChangeListener(null);
+  let clients = [];
+  let clientsPerBundleEntry = {};
+
+  let platformCache = {
```

Comment:
> I thought about merging the bundle entry and the platform, but I was unsure of whether the platform was relevant for something beyond the bundle entry, in another part of the packager

I was suggesting doing that only on this file which is pretty isolated from the rest of the packager.

> Also, do we really want a cache for each client?

On your PR we're having different caches for different combinations of bundleEntry + platform, which represents different clients except rare case on which you have both the simulator and a device connected to the Packager. We could easily make the client class support multiple web socket connections to support this use case as well. I think introducing the client class will make the code simpler because:
  1. State (the cache) will be collocated with behavior (`processFileChange`) which will make this code a bit more object oriented.
  2. You'll have to deal with the cache for the client instead of with a cache for multiple clients.
  3. Sending updates will be per client so you won't have to maintain an object with the updates you have to send.
  4. Error handling *per client* will be easy to implement.
  5. All this code will become much easier to test in isolation.

What do you think? :)